### ISSUE TITLE
[ App CALCULS ] ajout du modulo

### ISSUE THREAD

**Issue by fmOOmf** [UpsilonNumworks/Upsilon#193]
Bonjour à toute l'équipe 

J'ai peut être mal cherché mais je n'ai pas trouvé la fonction modulo dans l'application `Calculs.`
Il existe la fonction  REM pour les entiers mais elle n'accepte pas les réels.
Or en particulier pour la trigonométrie, il peut être très utile de trouver l'angle "à 2 pi près".
Exemple : -5 pi/3  donne pi/3.

Suggestion : introduire la fonction mod (modulo) qui accepte les nombres reels.
Syntaxe proposée:
` mod(x, y)` calculerait x modulo y au sens mathématique.
`mod(x, y, p)` permettrait de choisir l'intervalle du résultat.
Le 3e paramètre p permettrait d'indiquer que l'on souhaite : 
- le résultat dans `[ 0 y [ ` (p=0 : valeur par défaut) 
- le résultat dans `] -y/2  y/2 ]` (p=1).

Ex1 : pour avoir le modulo sur [ 0  2pi [  : 
`mod( -7pi/3 , 2pi, 0)`  => 5pi/3
Ou 
`mod( -7pi/3 , 2pi)`  => 5pi/3   

Ex2 : Pour avoir le résultat entre ] -pi. pi ] (choix par défaut de Upsilon pour le calcul des arguments complexes)
`mod( -7pi/3 , 2pi , 1)`  => - pi/3

Nb : 
 si x et y sont des entiers avec (x > y), alors
mod(x, y)  = rem(x, y)

**Comment by lolocomotive** [comment]
Il y a un opérateur modulo de mémoire (%)

**Comment by Lauryy06** [comment]
Le modulo en lui même ne doit pas être compliqué à implémenter (même s'il ne pourra pas simplifier des expressions comprenant des inconnues), mais je ne comprends pas vraiment le troisième paramètre. Il ne peut prendre comme valeur que 0 ou 1 ?

**Comment by fmOOmf** [comment]
Bonjour
oui, c'est ça.
Ma proposition (mais vous pouvez faire différemment) : 
- si p = 0, alors la fonction se comporte comme le modulo mathématique. 
  C'est la valeur par défaut (p=0 s'il n'est pas passé en paramètre).
- Si p = 1, alors, on a le fonctionnement qui permet de trouver l'angle comme le fait la fonction argument  de Upsilon.

!! Attentions aux bornes des intervalles du résultat (inclus/exclus). 

Si vous trouvez une syntaxe plus simple pour les élèves, n'hésitez pas !
:-)

**Comment by fmOOmf** [comment]
PS : 
Le challenge sera de conserver des fractions de pi quand on sera en radians
Ex : 15pi/4 devra donner 7pi/4  pour p=0 et -pi/4 pour p=1
Après vous pouvez imaginer des résultats complemetaires comme pour cos et sin ...

**Comment by Lauryy06** [comment]
Je pense que c'est n'est pas très intuitif. A mon avis deux fonctions différentes, un modulo classique et un modulo "mesure principale" qui fonctionne dans ]−π;π] seraient plus adaptées.

**Comment by fmOOmf** [comment]
Oui, ce sera sans doute plus simple. Y a plus qu'à trouver les noms.
:-)

**Comment by Overengined** [comment]
On pourrait les appeler "modulo trigonométrique" et "modulo standard" ?

------
We could call it "trigonetrical modulo" and "standard modulo" ?

**Comment by fmOOmf** [comment]
Projetons-nous dans la peau du lycéen type. Qu'est-ce qui est le plus intuitif ?
**mod(a,b) donne le résultat dans [0 b[**.
Il fait mod(18,10) ça donne 8
Il fait mod (-7pi/3,2pi) ça donne 5pi/3  
mod pour modulo, ça semble intuitif => Go **mod** pour le modulo standard selon moi.

Maintenant, quel nom pour la mesure principale de l'angle : 
- mp (mesure principale) ?  
- modt (modulo trigo comme tu le proposes)  
- modp (modulo principal) ?  

L'avantage de modp ou modt c'est qu'on peut ré-utiliser la commande mod et la transformer en ajoutant une seule lettre ... et que par ordre alphabetique, il sera juste après mod.
J'ai un faible pour modp parce que le p rappelle bien "principal" de "mesure principale".

**Comment by Overengined** [comment]
J’aime l’idée de mettre mod() pour la fonction la plus susceptible d’être utilisée, c’est du bon design.
En revanche mettons nous de nouveau dans la peau d’un lycéen type comme tu le suggère avec beaucoup de perspicacité: je pense qu’il serait plus intéressant d’avoir le « t » de « trigonométrie » dans ma fonction, comme ça , paniqué dans un DS, suant à grosse goute et envisageant sérieusement le seppuku (oui bon j’exagère un peu, mais vous avez l’image), je saurais d’instinct à quoi j’ai affaire.

**Comment by fmOOmf** [comment]
Hello.
Je plussoie alors. 
mod(a,b) : donnerait le résultat dans [0 b[.
modt(a) : donnerait le résultat dans ] -pi pi]

Avis aux courageux développeurs  :-)
